File: 02-programacion-basica/07-cadenas-y-caracteres/scala/soluciones/RetoNormalizarNombre.scala
def normalizar(texto: String): String =
  texto.trim.toLowerCase.split("\\s+").filter(_.nonEmpty).map(_.capitalize).mkString(" ")

def contarVocales(texto: String): Int =
  val vocales = "aeiouáéíóú"
  // count conserva solo los caracteres que cumplen el predicado.
  texto.toLowerCase.count(vocales.contains(_))

@main def retoNormalizarNombre(): Unit =
  val nombre = normalizar("  aDRian ivAN pisco  ")
  println(nombre)
  println(s"Vocales: ${contarVocales(nombre)}")

File: 02-programacion-basica/07-cadenas-y-caracteres/java/soluciones/RetoNormalizarNombre.java
public class RetoNormalizarNombre {
    public static void main(String[] args) {
        String nombre = "  aDRian ivAN pisco  ";
        String normalizado = normalizar(nombre);
        System.out.println(normalizado);
        System.out.println("Vocales: " + contarVocales(normalizado));
    }

    static String normalizar(String texto) {
        String[] palabras = texto.trim().toLowerCase().split("\\s+");
        StringBuilder resultado = new StringBuilder();
        for (String palabra : palabras) {
            if (palabra.isEmpty()) continue;
            // Se capitaliza solo la primera letra de cada palabra.
            resultado.append(Character.toUpperCase(palabra.charAt(0)))
                    .append(palabra.substring(1)).append(' ');
        }
        return resultado.toString().trim();
    }

    static int contarVocales(String texto) {
        int total = 0;
        for (char c : texto.toLowerCase().toCharArray()) {
            if ("aeiouáéíóú".indexOf(c) >= 0) total++;
        }
        return total;
    }
}
